### PR TITLE
Add oneDNN support for Half  LSTM on CPU

### PR DESCRIPTION
**Pull request by CaoE**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132607

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal

### CODE REVIEW

File: test/test_mkldnn.py [hunk 1]
```diff
@@ -1469,24 +1469,33 @@ def _lstm_params_list(self):
...
 
-    @unittest.skipIf(IS_WINDOWS, "Limit support for bf16 path")
     def test_lstm(self):
         seed = 2023
         torch.manual_seed(seed)
 
         params_list = self._lstm_params_list()
         for dtype in types:
             bf16 = True if dtype == torch.bfloat16 and torch.ops.mkldnn._is_mkldnn_bf16_supported() else False
+            fp16 = True if dtype == torch.half and torch.ops.mkldnn._is_mkldnn_fp16_supported() else False
```
Comment (vadimkantorov) [inline test/test_mkldnn.py:1486]: could maybe be replaced simply by `fp16 = dtype == torch.half and torch.ops.mkldnn._is_mkldnn_fp16_supported()`?

File: test/test_mkldnn.py [hunk 2]
```diff
@@ -1469,24 +1469,33 @@ def _lstm_params_list(self):
         params_list = list(params_dict.values())
         return params_list
 
-    def _cast_dtype(self, input, bf16):
+    def _cast_dtype(self, input, bf16, fp16):
         if bf16:
             input = input.to(torch.bfloat16)
+        elif fp16:
+            input = input.to(torch.half)
```
Comment (yanbing-j) [inline test/test_mkldnn.py:1476]: Suggest to update to `_cast_dtype(self, input, dtype):`, specify dtype to bfloat16 and float16.

File: test/test_mkldnn.py [hunk 3]
```diff
@@ -1510,12 +1520,18 @@ def test_lstm(self):
...
                     with torch.backends.mkldnn.flags(enabled=False):
                         torch.manual_seed(seed)
-                        output1, (hn1, cn1) = self._cast_dtype(model1, bf16)(self._cast_dtype(input1, bf16),
-                                                                             (self._cast_dtype(h1, bf16),
-                                                                             self._cast_dtype(c1, bf16)))
+                        output1, (hn1, cn1) = self._cast_dtype(model1, bf16, fp16)(
+                            self._cast_dtype(input1, bf16, fp16),
+                            (
+                                self._cast_dtype(h1, bf16, fp16),
+                                self._cast_dtype(c1, bf16, fp16),
```
Comment (yanbing-j) [inline test/test_mkldnn.py:1532]: Update according to _cast_dtype API change.

File: test/test_mkldnn.py (hunk 2 above)
Comment (jgong5) [reply test/test_mkldnn.py:1476]: Originally we would unconditionally cast to bf16 when specified but now it is possible to bypass, sounds not right to me. Should assert on the support in mkldnn here?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: test/test_mkldnn.py (hunk 2 above)
Comment (CaoE) [reply test/test_mkldnn.py:1476]: Originally `_cast_dtype`  cast input to bf16 according to the flag `bf16` which is set by `dtype == torch.bfloat16 and torch.ops.mkldnn._is_mkldnn_bf16_supported()`. Now we add a option of fp16, so we want to use dtype to check if cast input to bf16/fp16.

Comment (jgong5) [reply test/test_mkldnn.py:1476]: What is the behavior now if mkldnn doesn't support bf16/fp16? Is there a behavior change from the original code? I guess my major concern is that we pass through things silently.

File: test/test_mkldnn.py (hunk 1 above)
Comment (CaoE) [reply test/test_mkldnn.py]: Thanks for your comments. Revised.

File: test/test_mkldnn.py (hunk 3 above)
Comment (CaoE) [reply test/test_mkldnn.py]: Thanks for your comments. Revised.

File: test/test_mkldnn.py (hunk 2 above)
Comment (jgong5) [reply test/test_mkldnn.py:1476]: Does the caller guarantee the `_is_mkldnn_bf16_supported()` returns `True` when the dtype is bfloat16 (same for fp16)? If so, should we use assertion here on `_is_mkldnn_bf16_supported()`?

Comment (CaoE) [reply test/test_mkldnn.py:1476]: When initially supporting BF16 LSTM, bf16 check was overlooked Added a check.
Revised `test_lstm`. Removed `_is_mkldnn_bf16_supported`.

**Comment by CaoE** [comment]
@yanbing-j @jgong5 Could you please review this PR again ? Thanks.

**Comment by CaoE** [comment]
@yanbing-j @jgong5 Could you please review this PR again ? Thanks.

**Review by jgong5 — APPROVED** [review]

**Review by peterbell10 — APPROVED** [review]

**Pull request closed without merging**